base/expr.jl

```diff
@@ -371,6 +371,157 @@ macro constprop(setting, ex)
+    - If the method returns, the results will always be egal.
+
+Note: This in particular implies that the return value of the method must be
+      immutable. Multiple allocations of mutable objects (even with identical
+      contents) are not egal.
+
+Note: The idempotency assertion is made world-arge wise. More formally, write
+      fₐ for the evaluation of `f` in world-age `a`, then we require:
+
+          ∀ a, x, y: x === y → fₐ(x) === fₐ(y)
+
+      However, for two world ages `a, b` s.t. `a != b`, we may have `fₐ(x) !== fₐ(y)``
+
+Note: A further implication is that idempontent functions may not make their
+      return value dependent on the state of the heap or any other global state
+      that is not constant for a given world age.
+
+Note: The idempontency includes all legal rewrites performed by the optimizizer.
+      For example, floating-point fastmath operations are not considered idempotent,
+      because the optimizer may rewrite them causing the output to not be idempotent,
+      even for the same world age (e.g. because one ran in the interpreter, while
+      the other was optimized).
+
+# `:effect_free`
+
+The `:effect_free` setting asserts that the method is free of externally semantically
+visible side effects. The following is an incomplete list of externally semantically
+visible side effects:
+
+ - Changing the value of a global variable.
+ - Mutating the heap (e.g. an array or mutable value), except as noted below
+ - Changing the method table (e.g. through calls to eval)
+ - File/Network/etc. I/O
+ - Task switching
+
+However, the following are explicitly not semantically visible, even if they
+may be observable:
+
+ - Memory allocations (both mutable and immutable)
+ - Elapsed time
+ - Garbage collection
+ - Heap mutations of objects whose lifetime does not exceed the method (i.e.
+   were allocated in the method and do not escape).
+ - The returned value (which is externally visible, but not a side effect)
+
+The rule of thumb here is that an externally visible side effect is anything
+that would affect the execution of the remainder of the program if the function
+were not executed.
+
+Note: The effect free assertion is made both for the method itself and any code
+      that is executed by the method. Keep in mind that the assertion must be
+      valid for all world ages and limit use of this assertion accordingly.
+
+# `:nothrow`
+
+The `:nothrow` settings asserts that this method does not terminate abnormally
+(i.e. will either always return a value or never return).
+
+Note: It is permissible for :nothrow annotated methods to make use of exception
+      handling internally as long as the exception is not rethrown out of the
+      method itself.
+
+Note: MethodErrors and similar exceptions count as abnormal termination.
+
+# `:terminates_globally`
+
+The `:terminates_globally` settings asserts that this method will eventually terminate
+(either normally or abnormally), i.e. does not infinite loop.
```

Comment:
Not sure about this one, `does not infinite loop` sounds odd to me, but I'm not a native speaker.
```suggestion
(either normally or abnormally), i.e. does not loop indefinitely.
```